Note viewing key permissioning system

thomas-waite opened the pull request:
## Summary
This PR implements a note viewing key permissioning system. 

It allows third parties to be granted access to a note, by the note owner adding an IES encrypted viewing key to the `metaData` of the note. 

A `noteAccess` mapping has been implemented to keep track of which Ethereum addresses have been granted view access to a note. Being granted approval gives that address permission to call `updateNoteMetaData()` - this allows the approved party to in turn grant view access to additional parties. 

The `noteAccess` mapping is based on the block timestamp. Specifically, it is a mapping of a `noteHash` to a mapping of `addressID` to `blockTimeStamp`:

```
    mapping(bytes32 => mapping(bytes32 => uint256)) public noteAccess;
``` 

`addressID` is `keccak256(approvedAddress, noteHash)`. 

In order to extract addresses from the `metaData` of a note, a `MetaDataUtils.sol` library has been written.

## Todo's
**1) Explicit permissioning step test**

The permissioning step in `ZkAssetBase.updateNoteMetaData()` needs explicitly testing. It's function is currently implied from querying the `noteAccess` mapping in certain scenarios - using a `ZkAssetTest` for which the permission requirement was removed. 

This was done due to the difficulty in creating a note - you need to know the public key - owned by a truffle account. Only truffle accounts can send transactions.

**2) Consider the implications of `updateNoteMetaDataI()` being called on `confidentialTotalMinted()` and `confidentialTotalBurned()` notes**

Consider the implications of these notes being passed and potential mitigations

<!--- Describe your changes in detail -->

## Testing instructions
Various test cases to check the permissioning flow have been added to `ZkAsset.js`. 
<!--- Please describe how reviewers can test your changes -->

thomas-waite committed:
feat(protocol): implement MetaDataUtils library

thomas-waite committed:
test(protocol): success tests to check correct address extraction

thomas-waite committed:
feat(protocol): implement initial noteAccess permissioning system
add success tests to check noteAccess mapping correctly updates

thomas-waite committed:
refactor(protocol): remove extractAddresses() function
this extra loop is unnecessary and expensive. It can be combined with a loop that already takes place in the ZkAsset

thomas-waite committed:
feat(protocol): implement blockTimeStamp based permissioning system

thomas-waite committed:
test(protocol): test blockTimeStamp permissioning system
permissioning based tests, move various metaData based constants to the same location, update imports

ArnSch commented on packages/protocol/contracts/ERC1724/base/ZkAssetBase.sol:
```diff
@@ -311,23 +317,73 @@ contract ZkAssetBase is IZkAsset, IAZTEC, LibEIP712 {
         ( uint8 status, , , address noteOwner ) = ace.getNote(address(this), noteHash);
-        require(status == 1, "only unspent notes can be approved");
 
+        bytes32 addressID = keccak256(abi.encodePacked(msg.sender, noteHash));
```
This strikes me as superfluous if you're doing a two dimensional mapping.

Either do a mapping from `keccak256(abi.encodePacked(msg.sender, noteHash)) => uint256` or a two dimensional mapping from `noteHash => address => uint256`. Both will result in a unique timestamp per note per address.

thomas-waite replied:
Good point, my bad. It should just be a mapping of `keccak256(abi.encodePacked(msg.sender, noteHash)) => uint256`

ArnSch commented on packages/protocol/contracts/interfaces/IZkAsset.sol:
```diff
@@ -15,7 +15,7 @@ contract IZkAsset {
         bool _canConvert
     );
-    event ApprovedAddress(address indexed addressApproved, bytes32 indexed noteHash);
+    event ApprovedAddress(address indexed addressApproved, bytes32 indexed noteHash, address contractAddress);
```
Should contractAddress be indexed to make finding all approved notes for a user for an asset easier?

thomas-waite replied:
Chatted with the extension team about this to decide which fields should be indexed and arrived at the above. Will double check with them and mention your point, makes sense

thomas-waite replied:
Just spoke to Joe. They didn't need `contractAddress` indexing and now also don't need this event - so removing the event altogether

ArnSch commented on packages/protocol/contracts/libs/MetaDataUtils.sol:
```diff
@@ -8,23 +8,35 @@ pragma solidity >=0.5.0 <= 0.6.0;
+        * 0x81 - 0xa1 : approved addresses offset
+        * 0xa1 - 0xc1 : encrypted view keys offset
+        * 0xc1 - 0xe1 : app data offset
+        * 0xe1 - L_addresses : approvedAddresses
```
Where is L_addresses set? I assume that it's the first word at 0xe1, but would be good to document.

thomas-waite replied:
`L_addresses` isn't so much as set, rather it's just the length in memory that the addresses take up. The first word at `0xe1` is the number of addresses.

The extension has a variety of `metaData` encoding functions, which take a user's `metaData` intended for a note, and encodes it according to this ABI outlined in `MetaDataUtils.sol` . Will add a comment explaining this

thomas-waite replied:
The first word at `0xe1` is the number of addresses. `L_addresses` isn't so much as set, rather it's just the length in memory that the addresses take up

ArnSch replied:
Yes, I more mean documenting that the first word at 0xe1 is in fact num of addresses, same for L_encryptedViewKeys

thomas-waite replied:
Gotcha - this is standard across our ABI encoding. The first word of any dynamic bytes array, within another dynamic bytes array, is the number of elements in that array. We should document and make this ABI encoding clear

ArnSch commented on packages/protocol/contracts/libs/MetaDataUtils.sol:
```diff
@@ -8,23 +8,35 @@ pragma solidity >=0.5.0 <= 0.6.0;
+        * 0xc1 - 0xe1 : app data offset
+        * 0xe1 - L_addresses : approvedAddresses
+        * (0xe1 + L_addresses) - (0xe1 + L_addresses + L_encryptedViewKeys) : encrypted view keys
+        * (0xe1 + L_addresses + L_encryptedViewKeys) - (0xe1 + L_addresses + L_encryptedViewKeys + L_appData) : appData
```
ditto for L_encryptedViewKeys

thomas-waite replied:
Same as above

ArnSch commented on packages/protocol/contracts/libs/MetaDataUtils.sol:
```diff
@@ -8,23 +8,35 @@ pragma solidity >=0.5.0 <= 0.6.0;
+                add(
+                    add(
+                        metaData,
+                        add(0xe1, 0x20)  // go to the start of addresses, jump over first word
```
I think it's worth the overhead to sense check that you're not going out of the bounds of `0xe1 + L_addresses`. Could be as simple as checking that addressPos < L_addresses.

thomas-waite replied:
Sure, good idea - done

ArnSch commented on packages/protocol/test/ERC1724/ZkAsset.js:
```diff
@@ -330,7 +334,7 @@ contract('ZkAsset', (accounts) => {
-        it('should update the metadata of a note when ZkAsset.updateNoteMetadData() called', async () => {
+        it('should update the metadata of a note when ZkAsset.updateNoteMetaData() called', async () => {
             /**
              * Using ZkAssetTest as the ZkAsset here. ZkAssetTest has the safety feature
```
I don't feel comfortable with this solution. You can get the accounts used in ganache from a set mnemonic (see @charlie-g-cowan's branch).

This isn't really testing the zkAsset implementation of updateNoteMetadata. As it stands, we could drastically change that method implementation in zkAssetBase and this test would still pass.

thomas-waite replied:
Fully agree, the current testing setup is not good enough. This is one of the todo's I mentioned for the WIP PR - was implementing after this review

ArnSch commented on packages/protocol/test/ERC1724/ZkAsset.js:
```diff
@@ -343,43 +347,57 @@ contract('ZkAsset', (accounts) => {
+
+        it('should approve and grant addresses noteAccess', async () => {
             // create one outputNote, then place other addresses on the approved noteAccess mapping
             const zkAssetTest = await ZkAssetTest.new(ace.address, erc20.address, scalingFactor);
```
Ditto on zkAssetTest

thomas-waite replied:
Same as above

ArnSch commented on packages/protocol/contracts/test/ERC1724/ZkAssetTest.sol:
```diff
@@ -20,22 +20,28 @@ contract ZkAssetTest is ZkAsset {
         ( uint8 status, , , address noteOwner ) = ace.getNote(address(this), noteHash);
-        require(status == 1, "only unspent notes can be approved");
+        (, , bytes32 confidentialTotalMinted,
+        bytes32 confidentialTotalBurned, , , , bool canAdjustSupply) = ace.getRegistry(address(this));
```
```suggestion
        (
            ,,
            bytes32 confidentialTotalMinted,
            bytes32 confidentialTotalBurned
            ,,,,
            bool canAdjustSupply
        ) = ace.getRegistry(address(this));
```

thomas-waite replied:
Done

thomas-waite committed:
refactor(protocol): implement PR feedback
fix superfluous mapping, add some explanatory comments, update tests

thomas-waite committed:
feat(protocol): set deterministic ganache accounts
credit to Charlie Cowan - @charlie-g-cowan - work taken feat-wallet-develop branch

thomas-waite committed:
refactor(protocol): implement PR feedback
explicit check in MetaDataUtils.sol that addressPos is in bounds
fix circleci config.yml file, update tests

Commit:
Merge branch 'develop' into feat-noteAccess

ArnSch reviewed (approved):
LGTM
